crates/ruff_python_semantic/src/red_knot/semantic_index.rs

```diff
@@ -0,0 +1,652 @@
+use std::iter::FusedIterator;
+use std::sync::Arc;
+
+use rustc_hash::FxHashMap;
+
+use ruff_db::parsed::parsed_module;
+use ruff_db::vfs::VfsFile;
+use ruff_index::{IndexSlice, IndexVec};
+use ruff_python_ast as ast;
+
+use crate::red_knot::node_key::NodeKey;
+use crate::red_knot::semantic_index::ast_ids::AstIds;
+use crate::red_knot::semantic_index::builder::SemanticIndexBuilder;
+use crate::red_knot::semantic_index::symbol::{
+    GlobalScope, GlobalSymbol, LocalSymbolId, Scope, ScopeId, ScopesMap, SymbolTable,
+};
+use crate::Db;
+
+pub mod ast_ids;
+mod builder;
+pub mod definition;
+pub mod symbol;
+
+type SymbolMap = hashbrown::HashMap<LocalSymbolId, (), ()>;
+
+/// Returns the semantic index for `file`.
+///
+/// Prefer using [`symbol_table`] when working with symbols from a single scope.
+#[salsa::tracked(return_ref, no_eq)]
+pub(crate) fn semantic_index(db: &dyn Db, file: VfsFile) -> SemanticIndex {
+    let parsed = parsed_module(db.upcast(), file);
+
+    SemanticIndexBuilder::new(parsed).build()
+}
+
+/// Returns the symbol table for a specific `scope`.
+///
+/// Using [`symbol_table`] over [`semantic_index`] has the advantage that
+/// Salsa can avoid invalidating dependent queries if this scope's symbol table
+/// is unchanged.
+#[salsa::tracked]
+pub(crate) fn symbol_table(db: &dyn Db, scope: GlobalScope) -> Arc<SymbolTable> {
+    let index = semantic_index(db, scope.file(db));
+
+    index.symbol_table(scope.scope_id(db))
+}
+
+/// Returns a mapping from [`ScopeId`] to globally unique [`GlobalScope`].
```

Comment:
The operation is simple but it runs on every `public_symbol_ty`. The other reason is that `GlobalScope` is a salsa tracked struct and their identity is purely defined by their instance. That means, re-creating them would create different salsa `GlobalScope`s even if their value is identical. 

I think we could work around this by changing `GlobalScope` to 

```rust
struct GlobalScope {
	#[id]
	id: (VfsFile, ScopeId) 
}
```

But this shouldn't be necessary if we have this intermediate query, because Salsa than tracks the identity by their ordering.

Reply:
To me, having the global `Scope` have identity defined by value, and created on-demand, (i.e. basically what you propose using `#[id]`) seems simpler and cheaper than maintaining an extra query and redundant map structure. But it's quite likely that I'm missing some subtleties here.

Reply:
Let me try this tomorrow. I need to take a closer look at how salsa interns the tracked struct. Creating them lazily could especially be interesting for symbols because we would only create symbols for file symbols that are imported somewhere

Reply:
Okay I tried it out and it doesn't work. I added a log statement where we create the `ScopeId` from a `FileScopeId` and what I see is that salsa creates different IDs even when the two `ScopeIds` are equal

```
[crates/red_knot_python_semantic/src/semantic_index/symbol.rs:180:9] self = FileScopeId(
    0,
)
[crates/red_knot_python_semantic/src/semantic_index/symbol.rs:180:9] file = VfsFile(
    Id {
        value: 1,
    },
)
[crates/red_knot_python_semantic/src/semantic_index/symbol.rs:182:9] scope_id.debug(db) = ScopeId {
    [salsa id]: 0,
    file: VfsFile(
        Id {
            value: 1,
        },
    ),
    file_id: FileScopeId(
        0,
    ),
}
[crates/red_knot_python_semantic/src/semantic_index/symbol.rs:180:9] self = FileScopeId(
    0,
)
[crates/red_knot_python_semantic/src/semantic_index/symbol.rs:180:9] file = VfsFile(
    Id {
        value: 1,
    },
)
[crates/red_knot_python_semantic/src/semantic_index/symbol.rs:182:9] scope_id.debug(db) = ScopeId {
    [salsa id]: 1,
    file: VfsFile(
        Id {
            value: 1,
        },
    ),
    file_id: FileScopeId(
        0,
    ),
}
```

That means, it's our responsibility to track the `FileScopeId` -> `ScopeId` mapping

Reply:
Interesting. Does that mean that `#[id]` annotation doesn't do what we think it does? What's the explanation for this behavior? Is it not possible to have ingredients interned in salsa by value?

Reply:
You can intern values by id. We do that for module names. But interning has the downside that salsa never collects old values, at least in salsa 2022. 

What I understand is that salsa uses the id to match identical items across revisions (similar to key in React). But the uniqueness is not just defined by the id, but also the query that generates it. That means we could have a salsa query that maps the file id to the globally unique id, but that might result in more overhead (maybe not for public symbols if only a small subset is queried).